File: fridge.cpp
#include <iostream>
#include <string>
#include <vector>

using std::string;
using std::vector;
using std::cout;
using std::cin;
using std::endl;


struct ShoppingItem{
    string name;
    int amount;
    ShoppingItem(){
        this->name = "__NONAME";
        this->amount = 0;
    }
    ShoppingItem(string init_name){
        this->name = init_name;
        this->amount = 0;
    }
};

class FridgeDrawer{
public:
    string name;
    int limit;
    vector<ShoppingItem> contents;
    ShoppingItem current_item;
    int *current_amount;

    FridgeDrawer(){
        this->name = "__NONAME";
    }

    FridgeDrawer(string initname, int initlimit, vector<ShoppingItem> vecItems){
        this->name = initname;
        this->limit = initlimit;
        this->contents = vecItems;
    }

    void buyDrawer(){
        string buy_what;
        cout << "Product (";
        for(auto item: contents) cout << item.name << ": " << item.amount << ", ";
        cout << "): ";
        cin >> buy_what;

        if(!ValidItem(buy_what)) return;
        SelectItem(buy_what);
        cout << current_item.name << endl;
        cout << current_item.amount << endl;

        int buy_how_much;
        cout << "How much do you want to buy: ";
        cin >> buy_how_much;


        if((current_item.amount + buy_how_much) > (this->limit)){
            cout << "Drawer doesn't have that much space\n";
            return;
        } else {
            (*current_amount) += buy_how_much;
        }

    }

    void show_contents(){
        cout << "(";
        for(ShoppingItem item: contents)
            cout << item.name << ": " << item.amount << ", ";
        cout << ")";
    }

    void SelectItem(string item_name){
        for(auto item: contents){
            if(item_name==item.name){
                current_item = item;
                current_amount = &(item.amount);
                return;
            }
        }
    }

    bool ValidItem(string item_name){
            for(ShoppingItem item: contents){
                if(item_name==item.name){
                    return true;
            }
        }
        cout << "You can't put that item in this drawer.\n";
        return false;
    }
};




class Fridge{
private:
    vector<FridgeDrawer> drawers;
    FridgeDrawer current_drawer;
public:


    Fridge(vector<FridgeDrawer> initdrawers){
        this->drawers = initdrawers;
    }

    void buy_stuff(){
        string buy_more = "Yes";
        while(buy_more != "No"){
            string buy_what_kind;
            cout << "Product category (";
            for(auto drawer: drawers) cout << drawer.name << ", ";
            cout << "): ";
            cin >> buy_what_kind;

            if(!ValidDrawer(buy_what_kind)) continue;
            SelectDrawer(buy_what_kind);

            current_drawer.buyDrawer();

            string buy_more;
            cout << "Buy more? (Yes/No): ";
            cin >> buy_more;
        }
    }

    void show_stuff(){
        for(auto item: this->drawers){
            cout << item.name << ":\n";
            item.show_contents();
        }
    }

    bool ValidDrawer(string drawer_name){
        for(FridgeDrawer drawer: drawers){
            if(drawer_name==drawer.name){
                return true;
            }
        }
        cout << "This fridge does not have that drawer.\n";
        return false;
    }

    void SelectDrawer(string drawer_name){
        for(FridgeDrawer drawer: drawers){
            if(drawer_name==drawer.name){
                current_drawer = drawer;
                return;
            }
        }
    }
};


int main(){
    ShoppingItem coke("Coke");
    ShoppingItem water("Water");
    ShoppingItem club_mate("Mate");
    vector<ShoppingItem> vecBeverages = {coke, water, club_mate};

    FridgeDrawer beverages("Beverages", 10, vecBeverages);
    vector<FridgeDrawer> drawers = {beverages};

    Fridge myfridge(drawers);
    myfridge.buy_stuff();

    return 0;
}






// bool inIntVector(int item, vector<int> vec){
//     for(auto pos = vec.begin(); pos != vec.end(); ++pos){
//         if (*pos == item) return true;
//     }
//     return false;
// }

// void printStrVector(vector<string> vec){
//     cout << "{";
//     for(auto item: vec)
//         cout << item << ", ";
//     cout << "}\n";
// }

File: fridge.rb
class Fridge

    def initialize(items)
        @items = items
    end

    def buy_things()
        loop do
            puts "Do you want to buy more(Yes/No): "
            keep_buying = gets.chomp
            break unless keep_buying=="Yes"

            puts "What kind of stuff do you want: "
            @items.each_key {|key| print key.to_s + ", "}
            print ": "
            input_category = gets.chomp.to_sym
            if !(@items.key?(input_category))
                puts "That category doesn't exist"
                next
            end

            puts "What exactly do yo want: "
            @items[input_category].each_pair {|key, value| print key.to_s + ": " + value.to_s + ", " unless key=="LIMIT"}
            print ": "            
            input_product = gets.chomp.to_sym
            if !(@items[input_category].key?(input_product))
                puts "That product doesn't exist"
                next
            end

            puts "How much do you want to buy: "
            input_amout = gets.chomp.to_i
            if (@items[input_category][input_product] + input_amout) > @items[input_category]["LIMIT"]
                puts "The fridge will fit more than #{@items[input_category]["LIMIT"]} #{@items[input_category].to_s}"
                next
            else
                @items[input_category][input_product] += input_amout
            end

        end
        puts "You have: "
        puts
        show_items
    end

    def show_items()
        @items.each_key do |key|
            puts "#{key.to_s}: "
            @items[key].each_pair do |inner_key, inner_value|
                print "#{inner_key.to_s}: #{inner_value}, " unless inner_key=="LIMIT"
            end
            puts
        end
    end


end


# Buy what kind of stuff:
# Buy what exactly:
# Buy how much:

myfridge_init = {Beverages: {Coke: 0,
                             Mate: 0,
                             Water: 0,
                             Beer: 0,
                             "LIMIT" => 10},
                 Sweets: {GummyBears: 0,
                          Chocolate: 0,
                          ChewingGum: 0,
                          "LIMIT" => 10}}

myfridge = Fridge.new(myfridge_init)
myfridge.buy_things

